File: app/scripts/src/main/java/org/bitcoins/scripts/bitmex/BitmexProof.java
package org.bitcoins.scripts.bitmex;

import java.util.List;

public class BitmexProof {

  private long height = 0;
  private String blockhash;
  private String chain;
  private Long total;

  private List<String> keys;

  private BitmexClaim claim;

  private String xpub;


  public void setHeight(Long h) {
    height = h;
  }

  public Long getHeight() {
    return height;
  }

  public void setBlockhash(String hash) {
    blockhash = hash;
  }

  public String getBlockhash() {
    return blockhash;
  }

  public void setChain(String c) {
    chain = c;
  }

  public String getChain() {
    return chain;
  }

  public void setTotal(Long t) {
    total = t;
  }

  public Long getTotal() {
    return total;
  }

  public void setKeys(List<String> k)  {
    keys = k;
  }

  public List<String> getKeys() {
    return keys;
  }

  public void setXpub(String pub) {
    xpub = pub;
  }

  public String getXpub() {
    return xpub;
  }

  public void setClaim(BitmexClaim c) {
    claim = c;
  }

  public BitmexClaim getClaim() {
    return claim;
  }
}

File: app/scripts/src/main/scala/org/bitcoins/scripts/bitmex/BitmexProof.scala
package org.bitcoins.scripts.bitmex

import org.bitcoins.core.config.{NetworkParameters, Networks}
import org.bitcoins.core.crypto.ExtPublicKey
import org.bitcoins.core.currency.Satoshis
import org.bitcoins.crypto.DoubleSha256DigestBE

import scala.jdk.CollectionConverters.CollectionHasAsScala

case class BitmexProof(JBitmexProof: JBitmexProof) {
  val height: Long = JBitmexProof.getHeight

  val blockHash: DoubleSha256DigestBE =
    DoubleSha256DigestBE.fromHex(JBitmexProof.getBlockhash)
  val chain: NetworkParameters = Networks.fromString(JBitmexProof.getChain)
  val total: Satoshis = Satoshis(JBitmexProof.getTotal)
  val claim: BitmexClaim = BitmexClaim(JBitmexProof.getClaim)

  val xpub: Vector[ExtPublicKey] = {
    JBitmexProof.getXpub.asScala.toVector
      .map(ExtPublicKey.fromString)
  }

  val addresses: Vector[BitmexUtxo] = {
    JBitmexProof.getAddress.asScala.toVector
      .map(BitmexUtxo(_))
  }

  override def toString: String = {
    s"BitmexProof($blockHash,$chain,$total,$claim,$xpub,addressCount=${addresses.length})"
  }
}
